Feature/add map

offvanHooijdonk opened the pull request.

offvanHooijdonk committed:
Adding placing markers on the map based on photos geo-point.
Adding loading thumbnails to marker popup for the photo.

offvanHooijdonk committed:
Adding placing markers on the map based on photos geo-point.
Adding loading thumbnails to marker popup for the photo.

offvanHooijdonk committed:
Adding long press on Map and saving photo with LatLong.
Improving Photos loading on configuration change (screen rotation).

offvanHooijdonk committed:
Adding BottomSheet instead of dialog for place info and Gallery/Camera buttons.
Implementing retrieval of Place info by lat/long.

offvanHooijdonk committed:
Adding current location retrieval.
Adding feature to follow current location with FAB or drop following.
Enable adding photos with FAB again.

offvanHooijdonk committed:
Adding menu Filter.

offvanHooijdonk committed:
Fixing Logout, which was causing Main Activity issue.
Making all services PerFeature scoped.

offvanHooijdonk committed:
Adding search popup stub.

offvanHooijdonk committed:
Adding saving marker open state and Location mode through screen rotation.

offvanHooijdonk committed:
Slight code rearrangement

NikitaLychkovski commented on core_ui/src/main/java/by/off/photomap/core/ui/BaseFragment.kt:
```diff
@@ -1,13 +1,44 @@
+    protected fun checkPermission(permission: String) = ContextCompat.checkSelfPermission(ctx, permission) == PackageManager.PERMISSION_GRANTED
+
+    protected fun requestPermission(permission: String, force: Boolean, rationale: () -> Unit, onResult: (Boolean) -> Unit) {
+        onPermissionResult = onResult
```
could be discussed

NikitaLychkovski commented on storage_parse_impl/src/main/java/by/off/photomap/storage/parse/impl/GeoPointServiceImpl.kt:
```diff
@@ -0,0 +1,42 @@
+            var placeInfo = ""
+            try {
+                val addressList = geoCoder.getFromLocation(lat, lon, 5)
+                addrs@ for (ad in addressList) {
```
Try to use Kotlin Collection methods for such operations. Like find, firstOrNull, filter, foreach

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/MainActivity.kt:
```diff
@@ -56,26 +68,38 @@ class MainActivity : BaseActivity() {
+        super.onOptionsItemSelected(item)
         when (item?.itemId) {
             R.id.item_log_out -> {
                 Snackbar.make(mainRoot, "Please wait while we are logging you off", Snackbar.LENGTH_INDEFINITE).show() // todo implement some dialog here
```
hardcoded string

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/MainActivity.kt:
```diff
@@ -56,26 +68,38 @@ class MainActivity : BaseActivity() {
                 finish() // TODO place SplashActivity to this module and navigate to it here
-            } else {
-                Snackbar.make(mainRoot, error.message ?: "Some errorMessage occurred", Snackbar.LENGTH_LONG).show()
+            } else if (response?.error != null) {
```
hardcoded string

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/MainActivity.kt:
```diff
@@ -56,26 +68,38 @@ class MainActivity : BaseActivity() {
-            } else {
-                Snackbar.make(mainRoot, error.message ?: "Some errorMessage occurred", Snackbar.LENGTH_LONG).show()
+            } else if (response?.error != null) {
+                Snackbar.make(mainRoot, response.error?.message ?: "Some errorMessage occurred", Snackbar.LENGTH_LONG).show()
```
hardcoded string

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/map/AddPhotoBottomSheet.kt:
```diff
@@ -0,0 +1,81 @@
+
+
+        val data = arguments?.getParcelable<LatLng>(ARG_GEO_POINT)
+        if (data != null) {
```
more kotlin way

```
arguments?.getParcelable<LatLng>(ARG_GEO_POINT)?.let { 
            latLong = it
        } ?: kotlin.run {
            // todo show something?
            dismiss()
            return
        }
```

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/map/AddPhotoBottomSheet.kt:
```diff
@@ -0,0 +1,81 @@
+
+        progressPlaceLoad.show()
+
+        txtLat.text = formatLatitude(latLong.latitude, ctx)
```
Better to avoid using lateinit property for latLong

NikitaLychkovski replied:
possible solution:

```
private val latLong: LatLng? by lazy {  arguments?.getParcelable<LatLng>(ARG_GEO_POINT) }
if (latLong == null) {
            // todo show something?
            dismiss()
            return
        }
        latLong?.let {
            txtLat.text = formatLatitude(it.latitude, ctx)
            txtLong.text = formatLongitude(it.longitude, ctx)
        }
```

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/map/MapFragment.kt:
```diff
@@ -1,41 +1,83 @@
     private var progressDialog: AlertDialog? = null
+    private var dialogAddPhoto: DialogFragment? = null
+    // helpers
+    private val callbacks = mutableMapOf<String, CallbackHolder>()
```
not sure about this solution

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/photo/PhotoViewEditActivity.kt:
```diff
@@ -183,4 +193,34 @@ class PhotoViewEditActivity : BaseActivity() {
         supportActionBar?.setDisplayHomeAsUpEnabled(true)
     }
+
+    class IntentBuilder(private val ctx: Context) {
```
looks little weird or overcomplicated. It okay to create multiple fun in Companion object in activity

NikitaLychkovski commented on view_main/src/main/res/layout/dialog_add_photo.xml:
```diff
@@ -0,0 +1,82 @@
+        tools:visibility="visible" />
+
+    <TextView
+        android:id="@+id/txtFromGallery"
```
maybe it is make sense to create style for this?

offvanHooijdonk committed:
Implementing search dialog.
Adding View Model for Add Photo Dialog.
Adopt App Photo Dialog to landscape mode.

offvanHooijdonk committed:
Adding a ViewModel for Timeline Item, to hold view logic there

offvanHooijdonk committed:
Refactoring Thumbnail loading completely.
Presenting a separate module to load Thumbnails.

offvanHooijdonk committed:
Adding ViewModel for search results.
Reworked saving search history in Preferences.
Implemented searching tags.

offvanHooijdonk committed:
Fixing Photo View screen issue when previous data was shown.
Introducing Single Live Data for it.

offvanHooijdonk committed:
Mapping Search View Model to layout.
Implementing removal old tags from DB when editing a photo

NikitaLychkovski commented on core_utils/src/main/java/by/off/photomap/core/utils/PrefHelper.kt:
```diff
@@ -3,14 +3,33 @@ package by.off.photomap.core.utils
+
+    fun addSearchHistoryEntry(ctx: Context, entry: String) {
+        val historyList = getSearchHistory(ctx).toMutableList()
+        if (!historyList.map { it.toLowerCase() }.contains(entry.toLowerCase())) { // todo if contains - still need to reorder items
```
Maybe can be simplified
historyList.find { it.equals(entry, true) } != null

NikitaLychkovski commented on core_utils/src/main/java/by/off/photomap/core/utils/PrefHelper.kt:
```diff
@@ -3,14 +3,33 @@ package by.off.photomap.core.utils
+        val historyList = getSearchHistory(ctx).toMutableList()
+        if (!historyList.map { it.toLowerCase() }.contains(entry.toLowerCase())) { // todo if contains - still need to reorder items
+            if (historyList.size >= MAX_HISTORY) {
+                historyList.removeAt(historyList.size - 1)
```
Maybe you mean historyList.removeAt(MAX_HISTORY - 1) ?

NikitaLychkovski commented on core_utils/src/main/java/by/off/photomap/core/utils/PrefHelper.kt:
```diff
@@ -3,14 +3,33 @@ package by.off.photomap.core.utils
+        list.reduceIndexed { index, acc, s -> acc + (if (index > 0) HISTORY_SEPARATOR else "") + s }
 
-    private fun getPref() = PreferenceManager.getDefaultSharedPreferences(context)
+    private fun getPref(ctx: Context) = PreferenceManager.getDefaultSharedPreferences(ctx)
```
I think better to fo it with extension

NikitaLychkovski commented on thumb_loader/src/main/java/by/off/photomap/util/thumbs/ThumbnailLoaderImpl.kt:
```diff
@@ -0,0 +1,156 @@
+    // region Cache
+    private fun cacheFile(photoId: String, filePath: String) {
+        CoroutineScope(Dispatchers.Default).launch {
+            if (cachedFiles.size > CACHE_LIMIT) cachedFiles.keys.take(CACHE_LIMIT / 2).forEach { key -> cachedFiles.remove(key) }
```
Check Map fun minusAssign

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/map/AddPhotoBottomSheet.kt:
```diff
@@ -0,0 +1,84 @@
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        PhotoScreenComponent.get(ctx).inject(this)
+        viewModel = BaseFragment.getViewModel(this, viewModelFactory, AddPhotoDialogViewModel::class.java)
```
You have view model AddPhotoDialogViewModel but you are using  BaseFragment.getViewModel?

NikitaLychkovski replied:
I got idea but it is really not recommended to do it

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/map/AddPhotoBottomSheet.kt:
```diff
@@ -0,0 +1,84 @@
+        }
+        latLong?.also {
+            val parent = parentFragment
+            if (parent is MapFragment) parent.onAddPhotoDialogCreated(this)
```
can be discussed

offvanHooijdonk committed:
Adding search highlight.
Adding Tags in FlexLayout on Photo Info page

offvanHooijdonk committed:
Small fixes and refactoring

offvanHooijdonk committed:
Search implementation improved

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/map/MapFragment.kt:
```diff
@@ -46,12 +87,49 @@ class MapFragment : BaseFragment(), MainActivity.ButtonPhotoListener, MainActivi
+    }
+
+    override fun onMapReady(gMap: GoogleMap?) {
+        gMap?.also {
```
For null check is more common way to use .let

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/map/MapFragment.kt:
```diff
@@ -62,16 +140,133 @@ class MapFragment : BaseFragment(), MainActivity.ButtonPhotoListener, MainActivi
+    private fun updateMarkers(photoList: List<PhotoInfo>) {
+        val gMap = googleMap
+        gMap?.let {
+            it.clear()
```
maybe you need to clear markers here as well?

NikitaLychkovski commented on view_main/src/main/java/by/off/photomap/presentation/ui/map/MapFragment.kt:
```diff
@@ -62,16 +140,133 @@ class MapFragment : BaseFragment(), MainActivity.ButtonPhotoListener, MainActivi
+                    moveCameraToCurrent(LatLng(loc.latitude, loc.longitude))
+                }
+            } else {
+                (activity as MainActivity).setNavigationButtonMode(false)
```
it is better to use safe as?
(activity as? MainActivity)?.setNavigationButtonMode(false)

offvanHooijdonk committed:
Adding Room to store Search history cache results.

offvanHooijdonk committed:
Live search improvements and UI fixes.

offvanHooijdonk committed:
Moving Splash/Login activity to view_main module

offvanHooijdonk committed:
Adding Landscape support for Photo Edit and Timeline

offvanHooijdonk committed:
Adopting Photo View to Landscape mode

offvanHooijdonk committed:
Login screen adopted to Landscape mode

offvanHooijdonk committed:
Adding Close button on Photo Edit in Landscape mode

offvanHooijdonk committed:
Moving values to resources

offvanHooijdonk committed:
Minor changes and TODOs closed

offvanHooijdonk committed:
Resolving TODOs: removing redundant and implementing others.

offvanHooijdonk committed:
Fixing translations

offvanHooijdonk committed:
Adding google_api file to ignore list
